CIP-1854/CIP-1854.md

```diff
@@ -0,0 +1,243 @@
+ 
+
+# Specification
+
+## HD Derivation
+
+We consider the following HD derivation paths according to [CIP-1852]:
+
+```
+m / purpose' / coin_type' / account_ix' / role / index
+```
+
+
+To associate multi-signature keys to a wallet, we reserve `role=3` and `role=4` as an extension of [CIP-1852]. Wallet softwares should:
+
+- Derive multisig key indexes sequentially, starting from 0 and up to 2^31-1
+- Prevent the creation of new multisig keys before past keys are seen in an on-chain script.
+- Yet always provide up-to 20 consecutive unused multisig keys.
+- Keys should follow the same derivation scheme as described in [CIP-1852] (i.e. Ed25519-BIP32).
+
+Allowed values of the `role` level (at the moment of writing this specification) are summarized in the table below:
+
+role | description
+---  | ---
+0    | Public payment keys
+1    | Internal payment keys (change)
+2    | Stake keys
+3    | Multi-signatures payment keys
+4    | Multi-signatures stake keys
+
+### Rationale
+
+- Multisig keys are scoped to accounts, which allows wallet's owners to separate their activity easily.
+
+- One or many keys can be created from a parent public key via soft-derivation. This allows participant to easily share their multisig keys to participate in a multisig script without having to fetch for their hardware device. The device is still required for signing.
+
+### Examples
+
+- `m/1852’/1815’/0’/3/0`
+- `m/1852’/1815’/0’/4/14`
+- `m/1852’/1815’/0’/4/42`
+- `m/1852’/1815’/0’/3/1337`
+
+
+## User-facing encoding
+
+Multi-signatures payment verification and signing keys (`role=3`) with chain code should be presented bech32-encoded using `addr_xvk` and `addr_xsk` prefixes respectively, as specified in [CIP-5]. When represented without chain-code, `addr_vk` and `addr_sk` should be used instead.
+
+Similarly we use `stake_xvk`, `stake_xsk`, `stake_vk` and `stake_sk` for multi-signatures stake verification and signing keys (`role=4`).
+
+### Examples
+
+```yaml=
+- base16: | 
+    179be1ac9e63abb5fe4666df03007b07
+    33a3b63cdd08cd3635b8d364e16aaf26
+    49856f2ba513786afdbe5c7a07565c02
+    9ba7a4290d20aa3c80ecc1835841ed78
+  bech32:
+     addr_xvk1tv4ny6mcp82n8vxnedhql805wvu8vdu6pety9d7uafy0hdjr20f0smyna3wfhrs2ats809jkw4uf7te4e2sfjm0uh03rtfcwy7rdvxqhmnewl
+```
+
+## Multisig Wallets
+
+### Templates
+
+To define a multisig wallet, participants must provide: 
+
+- A payment script template.
+- A delegation script template (possibly null).
+- All the cosigners role=3 and role=4 (4th level) public keys involved in templates. Alternatively, cosigners may also share their account public key (3rd level) directly. 
```

Comment:
It all makes sense to me except for this point. Since we're re-using `1852'` instead of defining a new purpose, sharing the account public key is not ideal because it would reveal your keys for non-multisig addresses (and keys for any other spec implemented in the future). It would be unfortunate if future specs had to assume the account public key is known by others.

That being said, sharing two keys (`role=3` and `role=4`) is not ideal either because it means in the future if we want multisig to support some new specification that requires a new chain, multisig participants would have to share this new key (instead of just deriving it from the account key)

So given this, I think we have three main options:

1) We make multisig work on a different derivation path than `1852'` so account key can be shared (reveals all multisigs you are part of, but still slightly better). The cost is now you need two different wallets in the UI for the same recovery phrase
2) Add a new level to the multisig so that 
`role=3`: `m/1852'/1815'/0'/3/0/0`
`role=3`: `m/1852'/1815'/0'/3/1/0`
But this comes with the same problem as the `change=3` proposal we had when originally debating how to do staking keys
3) Keep it simple and use the account public key as described and accept the tradeoff.

Reply:
> 1. We make multisig work on a different derivation path than 1852' so account key can be shared (reveals all multisigs you are part of, but still slightly better). The cost is now you need two different wallets in the UI for the same recovery phrase

This isn't totally excluded indeed although, I see the 1852' as a good choice and a sort of 'extensible' purpose; especially with the way the CIP has been written. 
The rationale for sharing the account key was that in principle, we would encourage wallets to use different accounts for shared activities. It would get quite messy otherwise. This would make sharing an account key 'safe' to share: the account would only contain multisig keys.

Reply:
Note: after discussion during the last CIP meeting, we agreed that it would make more sense, in particular w.r.t to further extensions of the role level, to use a different purpose (purpose=1854') for this type of wallet and stick to the same structure for the role as role=1852' wallets (with however, an unused so-called internal chain role=1).

Reply:
Proposal amended as discussed :+1: